Support for 32-bit architectures

go-mysql uses github.com/pingcap/tidb/pkg/parser and this doesn't build on 32-bit architecures.

Note that fixing some of these issues would be good anyway. But the focus should be the parser. And I don't think this is ever going to be a supported architecture for TiDB so the priority is low.

> Hello, I'm one that cares :) @xhebox 
> [go-mysql](https://github.com/go-mysql-org/go-mysql), which has tidb as a dependency, fails to build on 32bit archs.
> 
> Here's one of our [goreleaser failed actions](https://github.com/conduitio-labs/conduit-connector-mysql/actions/runs/13312008116/job/37176673456) as an example.
> 
> We'll build for 64bit archs for now, not really a blocker, but would be nice to support 32bit. 

 _Originally posted by @alarbada in [#3227](https://github.com/pingcap/tidb/issues/3227#issuecomment-2679004365)_

```
dvaneeden@dve-carbon:~/dev/go-mysql$ GOARM=7 GOARCH=arm make build
go build -o bin/go-mysqlbinlog cmd/go-mysqlbinlog/main.go
go build -o bin/go-mysqldump cmd/go-mysqldump/main.go
go build -o bin/go-canal cmd/go-canal/main.go
# github.com/pingcap/tidb/pkg/parser
../../go/pkg/mod/github.com/pingcap/tidb/pkg/parser@v0.0.0-20241118164214-4f047be191be/hintparser.go:1359:69: cannot use math.MaxInt64 (untyped int constant 9223372036854775807) as int value in argument to ErrWarnMemoryQuotaOverflow.GenWithStackByArgs (overflows)
make: *** [Makefile:10: build] Error 1
dvaneeden@dve-carbon:~/dev/go-mysql$ git describe 
v1.7.0-212-g5942cc6
```

And with TiDB directly:
```
dvaneeden@dve-carbon:~/dev/pingcap/tidb$ git diff -U0
diff --git a/Makefile b/Makefile
index 1766194156..46793d548d 100644
--- a/Makefile
+++ b/Makefile
@@ -207 +207 @@ ifeq ($(TARGET), "")
-       CGO_ENABLED=1 $(GOBUILD) $(RACE_FLAG) -ldflags '$(LDFLAGS) $(CHECK_FLAG)' -o bin/tidb-server ./cmd/tidb-server
+       CGO_ENABLED=0 $(GOBUILD) $(RACE_FLAG) -ldflags '$(LDFLAGS) $(CHECK_FLAG)' -o bin/tidb-server ./cmd/tidb-server
dvaneeden@dve-carbon:~/dev/pingcap/tidb$ GOARM=7 GOARCH=arm CGO_ENABLED=0 make server
CGO_ENABLED=0  GO111MODULE=on go build -tags codes  -ldflags '-X "github.com/pingcap/tidb/pkg/parser/mysql.TiDBReleaseVersion=v9.0.0-alpha-320-g03f4a2e802-dirty" -X "github.com/pingcap/tidb/pkg/util/versioninfo.TiDBBuildTS=2025-02-25 09:07:04" -X "github.com/pingcap/tidb/pkg/util/versioninfo.TiDBGitHash=03f4a2e802e03f623c0b59b4fb4747b2495b5792" -X "github.com/pingcap/tidb/pkg/util/versioninfo.TiDBGitBranch=master" -X "github.com/pingcap/tidb/pkg/util/versioninfo.TiDBEdition=Community" ' -o bin/tidb-server ./cmd/tidb-server
# github.com/tikv/client-go/v2/internal/unionstore/arena
../../../go/pkg/mod/github.com/tikv/client-go/v2@v2.0.8-0.20250214064015-a805ea84f300/internal/unionstore/arena/arena.go:200:29: alignMask (untyped int constant 4294967288) overflows int
# github.com/pingcap/tidb/pkg/store/mockstore/unistore/lockstore
pkg/store/mockstore/unistore/lockstore/arena.go:152:29: alignMask (untyped int constant 4294967288) overflows int
# github.com/pingcap/badger/table/sstable
../../../go/pkg/mod/github.com/pingcap/badger@v1.5.1-0.20241015064302-38533b6cbf8d/table/sstable/table.go:542:17: math.MaxUint32 (untyped int constant 4294967295) overflows int
# github.com/tikv/client-go/v2/internal/client
../../../go/pkg/mod/github.com/tikv/client-go/v2@v2.0.8-0.20250214064015-a805ea84f300/internal/client/client.go:80:22: cannot use math.MaxInt64 - 1 (untyped int constant 9223372036854775806) as int value in variable declaration (overflows)
# github.com/lestrrat-go/jwx/v2/jwt/openid
../../../go/pkg/mod/github.com/lestrrat-go/jwx/v2@v2.0.21/jwt/openid/birthdate.go:69:7: math.MaxInt64 (untyped int constant 9223372036854775807) overflows int
# github.com/pingcap/tidb/pkg/parser
pkg/parser/hintparser.go:1358:69: cannot use math.MaxInt64 (untyped int constant 9223372036854775807) as int value in argument to ErrWarnMemoryQuotaOverflow.GenWithStackByArgs (overflows)
make: *** [Makefile:207: server] Error 1
dvaneeden@dve-carbon:~/dev/pingcap/tidb$ 
```

The `CGO_ENABLED` change is needed as I'm building on linux/amd64 for linux/arm

--- a/pkg/parser/hintparser.go
+++ b/pkg/parser/hintparser.go
@@ -1355,7 +1355,7 @@ yynewstate:
 					QBName:   ast.NewCIStr(yyS[yypt-3].ident),
 				}
 			} else {
-				yylex.AppendError(ErrWarnMemoryQuotaOverflow.GenWithStackByArgs(math.MaxInt64))
+				yylex.AppendError(ErrWarnMemoryQuotaOverflow.GenWithStackByArgs(math.MaxInt))
 				parser.lastErrorAsWarn()
 				parser.yyVAL.hint = nil
 			}

--- a/pkg/parser/hintparser.y
+++ b/pkg/parser/hintparser.y
@@ -316,7 +316,7 @@ TableOptimizerHintOpt:
 				QBName:   ast.NewCIStr($3),
 			}
 		} else {
-			yylex.AppendError(ErrWarnMemoryQuotaOverflow.GenWithStackByArgs(math.MaxInt64))
+			yylex.AppendError(ErrWarnMemoryQuotaOverflow.GenWithStackByArgs(math.MaxInt))
 			parser.lastErrorAsWarn()
 			$$ = nil
 		}

